app/controllers/AuthenticationController.scala

```diff
@@ -429,6 +431,91 @@ class AuthenticationController @Inject()(
+                           given_name: String,
+                           family_name: String,
+                           email: String) {
+    def username: String = preferred_username
+  }
+
+  object OpenConnectId {
+    implicit val format = Json.format[OpenConnectId]
+  }
+
+  def loginViaOpenIdConnect(): Action[AnyContent] = sil.UserAwareAction.async { implicit request =>
+    openIdClient.redirectUrl(oidcConfig, absoluteOidcCallbackURL).map(url => Redirect(url))
+  }
+
+  def loginUser(user: User, loginInfo: LoginInfo)(implicit request: Request[AnyContent]): Future[Result] =
+    userService.retrieve(loginInfo).flatMap {
+      case Some(user) if !user.isDeactivated =>
+        for {
+          authenticator: CombinedAuthenticator <- combinedAuthenticatorService.create(loginInfo)
+          value: Cookie <- combinedAuthenticatorService.init(authenticator)
+          result: AuthenticatorResult <- combinedAuthenticatorService.embed(value, Redirect("/dashboard"))
+          _ <- multiUserDAO.updateLastLoggedInIdentity(user._multiUser, user._id)(GlobalAccessContext)
+          _ = userDAO.updateLastActivity(user._id)(GlobalAccessContext)
+        } yield result
+      case None =>
+        Future.successful(BadRequest(Messages("error.noUser")))
+      case Some(_) => Future.successful(BadRequest(Messages("user.deactivated")))
+    }
+
+  // Is called after user was successfully authenticated
+  def loginOrSignupViaOidc(oidc: OpenConnectId): Request[AnyContent] => Future[Result] = {
+    implicit request: Request[AnyContent] =>
+      userService.userFromMultiUserEmail(oidc.email)(GlobalAccessContext).futureBox.flatMap {
+        case Full(user) =>
+          val loginInfo = LoginInfo("credentials", user._id.toString)
+          loginUser(user, loginInfo)
+        case Empty =>
+          for {
+            organization: Organization <- organizationService.findOneByInviteByNameOrDefault(None, None)(
+              GlobalAccessContext)
+            user: User <- userService.insert(organization._id,
+                                             oidc.email,
+                                             oidc.given_name,
+                                             oidc.family_name,
+                                             isActive = true,
+                                             userService.getOIDCPasswordInfo)
+            multiUser: MultiUser <- multiUserDAO.findOne(user._multiUser)(GlobalAccessContext)
+            _ = analyticsService.track(SignupEvent(user, hadInvite = false))
+            //  brainDBResult <- brainTracing.registerIfNeeded(user, signUpData.password).toFox // Is this necessary?
+            _ = if (conf.Features.isDemoInstance) {
+              mailchimpClient.registerUser(user, multiUser, tag = MailchimpTag.RegisteredAsUser)
+            } else {
+              Mailer ! Send(defaultMails.newUserMail(user.name, oidc.email, None, enableAutoVerify = true))
+            }
```

Comment:
I wonder if some of this function’s logic can be extracted into a method. A lot of this exists twice now, once for oidc and once for the normal registration/login

Reply:
done now